SPE checkout changes to match Woo visual standards (block checkout)

wjrosa opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes #3896

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

We are updating the Single Payment Element implementation to make all the payment methods it renders to look just like any other WooCommerce payment method to avoid any UX impact for shoppers. However, the Stripe container has some styling limitations.

This PR apply the required styles for the block checkout.

Preview for multiple payment methods:
<img width="895" alt="Screenshot 2025-03-18 at 08 58 10" src="https://github.com/user-attachments/assets/9d759abe-d207-4aa9-954d-5a25108b61a5" />

Preview when Stripe is the only available method:
<img width="902" alt="Screenshot 2025-03-18 at 09 32 53" src="https://github.com/user-attachments/assets/8067ee17-bc1a-4cfe-ade6-581c06979125" />

We will still collect some design feedback, which may result in additional changes (on other PRs).

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

- Checkout and build this branch on your test environment (`tweak/spe-changes-to-match-woo-visuals`)
- Connect your Stripe account
- On your Stripe dashboard, enable multiple payment methods
- Copy your child payment method configuration ID from here: `https://dashboard.stripe.com/settings/payment_methods` (labeled "Your configuration")
- Insert the ID above on `client/blocks/upe/upe-deferred-intent-creation/payment-elements.js:139`. This is temporary until the parent configuration, labeled WooCommerce Inc., supports all methods by default
- Insert the ID above on `client/classic/upe/payment-processing.js:136` as well
- Build the frontend files (`npm run build:webpack`)
- Enable the SPE feature flag (`_wcstripe_feature_spe`). You can do it by either hardcoding the return value of `is_spe_available` to `true` or by running `npm run wp option update _wcstripe_feature_spe 'yes'`
- As a merchant, disable the legacy checkout experience in settings (`wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings`)
- Enable the Single Payment Element feature
- I recommend setting your store currency to EUR so you can have more methods available
- As a shopper, add any product to your cart
- Go to the block checkout page
- Confirm you can see all the payment methods rendered just like the previews above
- Repeat the steps above after enabling another payment method (i.e., PayPal) and verify the it looks like the other preview above

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Tested on mobile (or does not apply)

### Changelog entry

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

wjrosa committed:
Initial work for SPE in the blocks checkout

wjrosa committed:
Initial work for SPE in the blocks checkout

wjrosa committed:
Merge branch 'develop' into add/spe-block-checkout-implementation3

wjrosa committed:
Including method to check if SPE is enabled

wjrosa committed:
Implementing the new method

wjrosa committed:
Implementing the new method

wjrosa committed:
Fix payment method types param

wjrosa committed:
Merge branch 'develop' into add/spe-block-checkout-implementation3

wjrosa committed:
Adding specific unit tests

wjrosa committed:
Merge branch 'add/spe-block-checkout-implementation3' of https://github.com/woocommerce/woocommerce-gateway-stripe into add/spe-block-checkout-implementation3

wjrosa committed:
Putting back tests removed by mistake

wjrosa committed:
Putting back tests removed by mistake

wjrosa committed:
Replacing payment method configuration ID

wjrosa committed:
Merge branch 'develop' into add/spe-block-checkout-implementation3

wjrosa committed:
Changelog and readme entries

wjrosa committed:
Merge branch 'add/spe-block-checkout-implementation3' of https://github.com/woocommerce/woocommerce-gateway-stripe into add/spe-block-checkout-implementation3

wjrosa committed:
Mninor change to PHP code coverage script

wjrosa committed:
Merge branch 'develop' into add/spe-block-checkout-implementation3

wjrosa committed:
Merge branch 'develop' into add/spe-block-checkout-implementation3

wjrosa committed:
SPE implementation for the classic/shortcode checkout

wjrosa committed:
Reverting hardcode of feature flag value

wjrosa committed:
Accordion layout

wjrosa committed:
Fix mandate and return URL parameters passing

wjrosa committed:
Change method title

wjrosa committed:
Merge branch 'develop' into add/spe-classic-checkout-implementation

wjrosa committed:
Fix feature flag case

wjrosa committed:
Fix payment method title

wjrosa committed:
Fix voucher payments processing

wjrosa committed:
Doc update + improving code readability

wjrosa committed:
Changelog and readme entries

wjrosa committed:
Fix tests

wjrosa committed:
Merge branch 'develop' into add/spe-classic-checkout-implementation

wjrosa committed:
Removing hardcoded payment method configuration ID

wjrosa committed:
Merge branch 'add/spe-classic-checkout-implementation' of https://github.com/woocommerce/woocommerce-gateway-stripe into add/spe-classic-checkout-implementation

wjrosa committed:
Merge branch 'develop' into add/spe-classic-checkout-implementation

wjrosa committed:
Fix payment method name for the pay for order page

wjrosa committed:
Improving readability

wjrosa committed:
SPE changes to match Woo visuals

wjrosa committed:
Changelog and readme entries

wjrosa committed:
Disabling saving of payment methods when SPE is enabled

wjrosa committed:
Disabling testing instructions + additional styling changes to SPE container

wjrosa committed:
Additional style changes

diegocurbelo committed:
Merge branch 'develop' into add/spe-classic-checkout-implementation

wjrosa committed:
Stripe icon

wjrosa committed:
Stripe icon

wjrosa committed:
Additional style changes

wjrosa committed:
Implementing the Stripe icon

diegocurbelo committed:
Merge branch 'develop' into add/spe-classic-checkout-implementation

wjrosa committed:
Additional style changes

wjrosa committed:
Merge branch 'develop' into add/spe-classic-checkout-implementation

wjrosa committed:
Merge branch 'develop' into add/spe-classic-checkout-implementation

wjrosa committed:
Merge branch 'add/spe-classic-checkout-implementation' into fix/order-pay-for-order-payment-method-name

wjrosa committed:
Merge branch 'fix/order-pay-for-order-payment-method-name' into tweak/spe-changes-to-match-woo-visuals

wjrosa committed:
Moving icon logic to a new method

wjrosa committed:
Improving styles, code readability and adding docs

wjrosa committed:
Simplifying the style applycation logic

wjrosa committed:
Changelog and readme updates

wjrosa committed:
Hiding the Stripe label when it is the only option available

wjrosa committed:
Removing hardcoded payment method configuration

wjrosa committed:
Adding specific unit tests

diegocurbelo committed:
Merge branch 'develop' into add/spe-classic-checkout-implementation

malithsen reviewed (changes requested):
Hi @wjrosa 
Nice work so far 👍 

I am getting the following fatal error when testing this PR

```
Fatal error: Uncaught Error: Call to a member function is_enabled() on null in /Users/malithsenaweera/projects/woocommerce-gateway-stripe/includes/payment-tokens/class-wc-stripe-payment-tokens.php:313
```

It's not obvious to me how your changes cause this but I only get this error on this branch.

I was able to uncomment the offending lines and proceed with testing. 
Noticed the following styling issues which we may want to address here, but all good if you prefer to do it on a separate PR(s).

1. On Block themes like twenty-twenty-five, the right border overflows the parent container. This doesn't happen on storefront themes.

![CleanShot 2025-03-18 at 15 51 10](https://github.com/user-attachments/assets/26e1576c-6c33-48a6-9a1c-885ff47bd180)

2. Stripe's iframe no longer uses the global stylings from the page. This looks like a regression as SPE on `develop` branch works correctly.

![CleanShot 2025-03-18 at 15 35 43](https://github.com/user-attachments/assets/d3159bde-3747-4c04-ba29-34c08d742760)

wjrosa committed:
Merge branch 'develop' into add/spe-classic-checkout-implementation

wjrosa committed:
Temporarily disabling the SEPA tokens for other methods feature when SPE is enabled

wjrosa committed:
Merge branch 'add/spe-classic-checkout-implementation' into fix/order-pay-for-order-payment-method-name

wjrosa committed:
Merge branch 'fix/order-pay-for-order-payment-method-name' into tweak/spe-changes-to-match-woo-visuals

wjrosa committed:
Fix style regression

malithsen reviewed (commented):
Thanks for fixing the Appearance API regression @wjrosa! I still see the border overflow in Blocks themes. 
Let me know if you prefer to fix it in a follow-up PR.

wjrosa committed:
Fix SPE container width when using different themes

wjrosa commented:
Good catches, @malithsen !

> I am getting the following fatal error when testing this PR

This issue was happening when "SEPA tokens for other methods" was enabled. I have fixed it on the base branch here https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4036#issuecomment-2736996895

> On Block themes like twenty-twenty-five, the right border overflows the parent container. This doesn't happen on storefront themes.

Fixed it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4053/commits/28b80def8fbb57ded30713c000fb6657cafc0df5

> Stripe's iframe no longer uses the global stylings from the page. This looks like a regression as SPE on develop branch works correctly.

I made a mistake when adding the custom rules to style the checkbox icons. Fixed it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4053/commits/9c5e4fbb101acb94a7469b20a1fb95769f79951c

malithsen reviewed (approved):
Looks great now. Nice work here!

wjrosa committed:
Removing unnecessary changes from the base branch to move it to develop

wjrosa committed:
Merge branch 'develop' into tweak/spe-changes-to-match-woo-visuals
